### PR TITLE
adding like terms

### PR DESCRIPTION
**Pull request by evykassirer**

### CODE REVIEW

File: lib/PolynomialTermOperations.js [hunk 1]
```diff
@@ -107,9 +112,46 @@ function addlikeTermPolynomialNodes(node) {
...
+  nodeStatus = addNegativeOneCoefficient(oldNode);
+  if (nodeStatus.hasChanged()) {
+    subSteps.push(nodeStatus);
+    // clone and reset
+    oldNode = NodeStatus.resetChangeGroups(nodeStatus.newNode);
+  }
+
+  // STEP 3: group the coefficients in a sum
+  let newNode = groupCoefficientsForAdding(clone(oldNode, false));
+  nodeStatus =  NodeStatus.nodeChanged(                                         // TODO - change groups better
```
Comment (evykassirer) [inline lib/PolynomialTermOperations.js]: these change groups should be only on the coefficients and coefficient sum

File: lib/PolynomialTermOperations.js [hunk 2]
```diff
@@ -119,12 +161,80 @@ function addlikeTermPolynomialNodes(node) {
...
+
+function addPositiveOneCoefficient(node) {                                      // TODO - change groups better
+  const oldNode = node;
+  const newNode = clone(node, false);
+  let change = false;
+
+  newNode.args = newNode.args.map(polyNode => {
+    const polyTerm = new PolynomialTermNode(polyNode);
+    if (polyTerm.getCoeffValue() === 1) {
+      change = true;
```
Comment (evykassirer) [inline lib/PolynomialTermOperations.js]: I want to set change groups on the old node and new node with new coefficient

File: lib/PolynomialTermOperations.js [hunk 3]
```diff
@@ -119,12 +161,80 @@ function addlikeTermPolynomialNodes(node) {
...
+  const oldNode = node;
+  const newNode = clone(node, false);
+  // the node is now always a * node with the left child the coefficent sum
+  // e.g. (2 + 4 + 5) and the right node the symbol part e.g. x or y^2
+  const coefficientSum = newNode.args[0];
+  coefficientSum.changeGroup = 1;
+  let newCoefficient = evaluateConstantSum(coefficientSum).newNode;             // TODO steps
+  newCoefficient.changeGroup = 1;
+  newNode.args[0] = newCoefficient;
   return NodeStatus.nodeChanged(
```
Comment (evykassirer) [inline lib/PolynomialTermOperations.js]: this should be a childChanged status

File: lib/PolynomialTermOperations.js [hunk 4]
```diff
@@ -107,9 +112,46 @@ function addlikeTermPolynomialNodes(node) {
...
+  if (nodeStatus.hasChanged()) {
+    subSteps.push(nodeStatus);
+    // clone and reset
+    oldNode = NodeStatus.resetChangeGroups(nodeStatus.newNode);
+  }
+
+  // STEP 3: group the coefficients in a sum
+  let newNode = groupCoefficientsForAdding(clone(oldNode, false));
+  nodeStatus =  NodeStatus.nodeChanged(                                         // TODO - change groups better
+    MathChangeTypes.ADD_POLYNOMIAL_TERMS, oldNode, newNode);
```
Comment (evykassirer) [inline lib/PolynomialTermOperations.js]: make a new change type for grouping coefficients

File: lib/evaluateArithmetic.js [hunk 5]
```diff
@@ -13,6 +13,15 @@ function evaluateArithmeticDFS(node) {
   let innerNodeStatus;
   if (NodeType.isParenthesis(node)) {
     innerNodeStatus = evaluateArithmeticDFS(node.content);
+    if (innerNodeStatus.hasChanged()) {
```
Comment (evykassirer) [inline lib/evaluateArithmetic.js]: don't have to do this anymore cause it should be handled before this function is called

File: test/PolynomialTerm.test.js [hunk 6]
```diff
@@ -36,13 +36,37 @@ describe('classifies symbol terms correctly', function() {
 function testCombinePolynomialTerms(exprStr, outputStr) {
   it(exprStr + ' -> ' + outputStr, function () {
     const inputNode = flatten(math.parse(exprStr));
-    const combinedNode = PolynomialTermOperations.combinePolynomialTerms(inputNode).newNode;
+    const combinedStatus = PolynomialTermOperations.combinePolynomialTerms(inputNode);
+    combinedStatus.subSteps.forEach(step => {
```
Comment (evykassirer) [inline test/PolynomialTerm.test.js]: get rid of these lines

File: lib/PolynomialTermOperations.js [hunk 7]
```diff
@@ -119,12 +161,80 @@ function addlikeTermPolynomialNodes(node) {
...
+  if (change) {
+    return NodeStatus.nodeChanged(
+        MathChangeTypes.ADD_COEFFICIENT_OF_ONE, oldNode, newNode);
+  }
+  else {
+    return NodeStatus.noChange(newNode);
+  }
+}
+
+function addNegativeOneCoefficient(node) {                                      // TODO - change groups better
```
Comment (evykassirer) [inline lib/PolynomialTermOperations.js]: todo: docstrings for new functions

**Comment by evykassirer** [comment]
ready to be looked at now, just adding tests

File: lib/evaluateConstantSum.js [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+// Returns a NodeStatus object.
+function evaluateConstantSum(node) {
+  if (NodeType.isParenthesis(node)) {
+    node = node.content;
+  }
+  if (!NodeType.isOperator(node) || node.op !== '+') {
+    return NodeStatus.noChange(node);
+  }
+
+  // functions needed to evalute the sum
```
Comment (evykassirer) [inline lib/evaluateConstantSum.js]: **evaluate

File: lib/PolynomialTermOperations.js [hunk 9]
```diff
@@ -107,24 +111,150 @@ function addlikeTermPolynomialNodes(node) {
   if (!canAddLikeTermPolynomialNodes(node)) {
     return NodeStatus.noChange(node);
   }
-  const polynomialTermList = node.args.map(n => new PolynomialTermNode(n));
 
-  // get a list of the coefficients, to be added
+  let oldNode = clone(node, false);
+  const subSteps = [];
+
+  // STEP 1: IF any nodes have no coefficient, make it have coefficient 1
```
Comment (aelnaiem) [inline lib/PolynomialTermOperations.js]: Why is the if all uppercase?

File: lib/evaluateConstantSum.js [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+  const summingFunctions = [
+    evaluateArithmetic,
+    ConstantFraction.addConstantFractions,
+    ConstantFraction.addConstantAndFraction,
```
Comment (aelnaiem) [inline lib/evaluateConstantSum.js]: Does addConstantAndFraction add constants to variable fractions or to constant fractions? Or both. This suggests both.

File: lib/PolynomialTermOperations.js (hunk 9 above)
Comment (evykassirer) [inline lib/PolynomialTermOperations.js]: to try to make it clear that that step doesn't always happen, but I don't think it's very effective :p I'll update it and step 2 as well

**Comment by aelnaiem** [comment]
Couple comments about naming, but otherwise looks good!! 👍

File: lib/evaluateConstantSum.js [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+        return status;
+      }
+    }
+  }
+
+  let oldNode = clone(node, false);
+  const subSteps = [];
+
+  // STEP 1: group fractions and constants separately
+  let status = groupConstantsAndFractions(oldNode);
```
Comment (evykassirer) [inline lib/evaluateConstantSum.js]: I use constant in most places in the code to refer to constant nodes, ie a single node with any number value (whole number or decimal number)

should I add this to the readme? should we rename things?

File: lib/evaluateConstantSum.js (hunk 10 above)
Comment (evykassirer) [inline lib/evaluateConstantSum.js]: both

File: lib/evaluateConstantSum.js (hunk 11 above)
Comment (aelnaiem) [inline lib/evaluateConstantSum.js]: ah I get it, that makes sense

File: lib/evaluateConstantSum.js (hunk 10 above)
Comment (aelnaiem) [inline lib/evaluateConstantSum.js]: cool!

**Pull request merged**